controllers/endpoint_controller.go

```diff
@@ -59,154 +58,195 @@ func (c *EndpointController) Reconcile(obj interface{}) error {
-			}
+		return c.handleDeletion(obj)
+	}
 
-			return nil
-		}
+	// Defer block to handle status updates for non-deletion paths
+	defer func() {
+		c.updateStatusOnError(obj, err)
+	}()
 
-		klog.Info("Deleting endpoint " + obj.Metadata.Name)
+	// always exec connect model to cluster, for cluster may dynamic scale, we need ensure model exists on all cluster nodes.
+	// todo: In order to reduce model connection actions, a new controller may be created in the future to uniformly manage model connections on the cluster.
+	err = c.connectModelToCluster(obj)
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect model %s to endpoint %s", obj.Spec.Model, obj.Metadata.Name)
+	}
 
-		err = c.gw.DeleteEndpoint(obj)
-		if err != nil {
-			return errors.Wrapf(err, "failed to delete route for endpoint %s", obj.Metadata.Name)
-		}
+	// Handle different phases
+	switch {
+	case obj.Status == nil || obj.Status.Phase == "" || obj.Status.Phase == v1.EndpointPhasePENDING:
+		return c.handlePendingPhase(obj)
+	case obj.Status.Phase == v1.EndpointPhaseFAILED:
+		return c.handleFailedPhase(obj)
+	case obj.Status.Phase == v1.EndpointPhaseRUNNING:
+		return c.handleRunningPhase(obj)
+	}
 
-		err = c.cleanupEndpoint(obj)
-		if err != nil {
-			return errors.Wrapf(err, "failed to cleanup endpoint %s", obj.Metadata.Name)
-		}
+	return nil
+}
 
-		err = c.disconnectModelFromCluster(obj)
-		if err != nil {
-			return errors.Wrapf(err, "failed to disconnect model %s to endpoint %s", obj.Spec.Model, obj.Metadata.Name)
-		}
+func (c *EndpointController) handleDeletion(obj *v1.Endpoint) error {
+	if obj.Status != nil && obj.Status.Phase == v1.EndpointPhaseDELETED {
+		klog.Infof("Endpoint %s already marked as deleted, removing from DB", obj.Metadata.Name)
 
-		// Update status to DELETED
-		err = c.updateStatus(obj, c.formatStatus(v1.EndpointPhaseDELETED, nil))
+		err := c.storage.DeleteEndpoint(strconv.Itoa(obj.ID))
 		if err != nil {
-			return errors.Wrapf(err, "failed to update endpoint %s status to DELETED", obj.Metadata.Name)
+			return errors.Wrapf(err, "failed to delete endpoint %s/%s from DB",
+				obj.Metadata.Workspace, obj.Metadata.Name)
 		}
 
 		return nil
 	}
 
-	// always exec connect model to cluster, for cluster may dynamic scale, we need ensure model exists on all cluster nodes.
-	// todo: In order to reduce model connection actions, a new controller may be created in the future to uniformly manage model connections on the cluster.
-	err = c.connectModelToCluster(obj)
-	if err != nil {
-		return errors.Wrapf(err, "failed to connect model %s to endpoint %s", obj.Spec.Model, obj.Metadata.Name)
+	klog.Info("Deleting endpoint " + obj.Metadata.Name)
+
+	// For deletion, we need to track if it succeeds to set correct phase
+	deleteErr := c.performDeletion(obj)
+
+	// Update status to DELETED if successful, or FAILED if not
+	phase := v1.EndpointPhaseDELETED
+	if deleteErr != nil {
+		phase = v1.EndpointPhaseFAILED
```

Comment:
Can we add a state to indicate a normal deletion, whether the deletion succeeded or failed, such as `Terminating`? Currently, the `Failed` state is used to indicate a normal coordination failure, which can be misleading.

Reply:
ok